src/conv/invokers/mlir_impl_gemm.cpp

```diff
@@ -59,16 +59,44 @@ struct MlirConvArgs
 // infer actual layout
 // - For NCHW, sizes = {N, C, H, W}; strides = {C*H*W, H*W, W, 1}
 // - For NHWC, sizes = {N, C, H, W}; strides = {C*H*W, 1, W*C, C}
-auto PermuteDimStride(const std::vector<size_t>& dims, const std::vector<size_t>& strides)
+void PermuteDimsStrides(std::vector<size_t>& dims, std::vector<size_t>& strides)
 {
     auto sorted_dims    = dims;
     auto sorted_strides = strides;
     auto p              = TensorDescriptor::find_permutation(dims, strides);
     std::transform(p.begin(), p.end(), sorted_dims.begin(), [&](auto i) { return dims[i]; });
     std::transform(p.begin(), p.end(), sorted_strides.begin(), [&](auto i) { return strides[i]; });
-    return std::make_tuple(sorted_dims, sorted_strides);
+    dims    = sorted_dims;
+    strides = sorted_strides;
 };
 
+void InsertGToDimsStrides(const std::string& layout,
+                          char dim,
+                          int group_count,
+                          std::vector<size_t>& dims,
+                          std::vector<size_t>& strides)
+{
+    auto dims_with_g    = dims;
+    auto strides_with_g = strides;
+    std::size_t index   = layout.find(dim);
+    if(index == std::string::npos)
+        MIOPEN_THROW("Failed to find channel in the layout");
+
+    // For dimensions,
+    //    Insert an additional dimension g before channel
+    //    Amend the channel size to be channel_size / group_count
+    dims_with_g[index] /= group_count;
```

Comment:
I think you can modify `dims` directly and get rid of `dims_with_g` together with two extra vector assignments.